dev_tools/utils/nxdl_utils.py

```diff
@@ -127,23 +127,21 @@ def get_nx_namefit(hdf_name, name, name_any=False):
     """
     if name == hdf_name:
         return len(name) * 2
+    if re.search(r"^[0-9]", hdf_name) is not None:
+        # Don't match anything if the name starts with a number
+        return -1
 
     uppercase_parts = re.findall("[A-Z]+(?:_[A-Z]+)*", name)
 
+    regex_name = name
     for up in uppercase_parts:
-        name = name.replace(up, r"([a-zA-Z0-9_]+)")
+        regex_name = regex_name.replace(up, r"[a-z0-9_]+")
 
-    name_match = re.search(rf"^{name}$", hdf_name)
+    name_match = re.search(rf"^{regex_name}$", hdf_name)
     if name_match is None:
         return 0 if name_any else -1
 
-    fit = 0
-    for up, low in zip(uppercase_parts, name_match.groups()):
-        for i in range(min(len(up), len(low))):
-            if up[i].lower() == low[i]:
-                fit += 1
-
-    return fit
+    return SequenceMatcher(None, hdf_name.upper(), name.upper()).ratio()
```

Comment:
Note that this operation is not commutative, so the opposite order of hdf_name and name might make more sense. Also note that matching needs to be recomputed, so quick_ratio might be a better choice.

Reply:
I wouldn't use `quick_ratio` as this just computes the ratio of matching characters regardless of their sequence in both strings. This is not what we want. But generally, I'm not exactly sure what we _do_ want here. The original implementation just counted the direct subsequent uppercase matches, i.e., 
- `testname` in `TESTname` got a score of 4
- but also `test` in `TESTname`
- similarly `TEST` in `TESTname`
- `MY_NAME_name` would have gotten a score of 0. 
- `TEEST` in `TESTname` would have gotten a score of 1.

On the order I agree, that it makes more sense the other way around.

Reply:
The whole thing is (only) about ambiguity resolution. Lower case letters shall match always (if nameType=any was not set) and they shall be also counted to be able to distinguish between e.g. SOURCE and sourceTYPE. If there are 2 concepts of the same type but (obviously) with different names, how can we decide which one a given instance with its hdf_name would belong to. It is not only about how to set up a metric which can be computed, but also that its computation must be trivial, so when a user choses a name for the hdf entry, it shall be evident which concept it will be belonging to. Note that this must also be considered by the person creating the definitions, too, to avoid any confusion and ambiguity.  
  
Old NeXus convention is to use name.lower() and maybe suffix it with a number, like for matching ENTRY one can use entry1, entry2, etc. This convention was implemented by providing a fit metric by counting the exact matching characters. Hence, if there are 2 concepts in a group with the same type and with (uppercase parts of the) names being close, like ENTITY(NXsomething) and ENTIRE(NXsomething), one should have started the corresponding hdf_names with entr... or etiti... at least for being able to assign them correctly.  
  
Indeed, one can argue, that a full match independent of the location could also be evident, like my_entity vs. my_entire. This can be supported SequenceMatcher, but both the definition designer and the data producer shall avoid situations, like 'my_entire_entity'. Similarly, to maintain the old convention which would unambiguously assign entity1213141523412 to ENTITY, it shall be avoided that this hdf_name gets higher ratio when compared to a concept like RUN12131415.

Reply:
LGTM

Reply:
I implemented it according to my suggestion now